CH6: Async programming (#146)
Web3NL/motoko-book: src/advanced-concepts/async-programming.md

**[Errors and Traps](#errors-and-traps)**  
[Errors](#errors)  
[Traps](#traps)

**[Async* and Await*](#async-and-await-1)**  
[`await` and `await*`](#await-and-await)  
[Atomic `await*`](#atomic-await)  
[Non-atomic `await*`](#non-atomic-await)  
[Keeping track of state commits and message sends](#keeping-track-of-state-commits-and-message-sends)

**[Table of asynchronous functions in Motoko](#table-of-asynchronous-functions-in-motoko)**

**[Try-Catch Expressions](#try-catch-expressions)**

## Async and Await

A call to a [shared function](/internet-computer-programming-concepts/actors.html#public-shared-functions-in-actors) immediately returns a _future_ of type [`async T`](/internet-computer-programming-concepts/actors.html#shared-async-types), where T is a [*shared type*](/internet-computer-programming-concepts/async-data/shared-types.html). A _future_ of type `async T` can be _awaited_ using the `await` keyword to retrieve the value of type `T`.

- Shared function calls and awaits are only allowed in [_asynchronous context_](#messaging-restrictions).
- Shared function calls that immediately return a future, do not suspend execution of the code.

> - Awaits using `await` do suspend execution of the function until a [_callback_](#shared-functions-that-await) is received.

Reviewer: ```suggestion
- Instead, execution proceeds immediately until the code chooses to `await` the result of the future.
- Awaiting a future suspends the computation until the future is complete and the waiter can make progress with its result.
```

(not sure about the link).

To actually retrieve the `Nat` value, we have to `await` the future. The actor sends a [message](#messages-and-atomicity) to itself with the request and halts execution until a response is received.

The result is a `Nat` that we increment within the [_callback_](#messages-and-atomicity) and use as the return value for `call_read`.

> **NOTE**  
> _Shared functions that `await` are not atomic. `call_read()` is executed as several separate [messages](#messages-and-atomicity), see [shared functions that `await`](#shared-functions-that-await)._

## Inter-Canister Calls

Actors can also call the shared functions of other actors. This always happens from within an [_asynchronous context_](#messaging-restrictions).

### Importing other actors

To call the shared functions of other actors, we need the [_actor type_](/internet-computer-programming-concepts/actors.html#actor-type) of the external actor and an _actor reference_.

```motoko
{{#include _mo/async-calls2.mo:a}}
```

> The actor type we use may be a [_subtype_](/advanced-types/subtyping.html) of the external actor type. We declare the actor type `actor {}` with only the shared functions we are interested in. In this case, we are importing the actor from the previous example and are only interested in the [query function](/internet-computer-programming-concepts/actors.html#public-shared-query) `read`.

Reviewer: ```suggestion
The actor type we use may be a [_supertype_](/advanced-types/subtyping.html) of the external actor type. We declare the actor type `actor {}` with only the shared functions we are interested in. In this case, we are importing the actor from the previous example and are only interested in the [query function](/internet-computer-programming-concepts/actors.html#public-shared-query) `read`.
```

Reviewer: It's the other way round.

## Messages and atomicity

From the [official docs](https://internetcomputer.org/docs/current/developer-docs/security/rust-canister-development-security-best-practices/#inter-canister-calls-and-rollbacks):  
_A message is a set of consecutive instructions that a subnet executes for a canister._

We will not cover the terms 'instruction' and 'subnet' in this book. Lets just remember that a single call to a [shared function](/internet-computer-programming-concepts/actors.html#public-shared-functions-in-actors) can be split up into several messages that execute separately.

A call to a shared function of any actor A, whether from an [_external client_](/internet-computer-programming-concepts/actors/canister-calling.html), [from itself](#async-and-await) or from another actor B (as an [Inter-Canister Call](#inter-canister-calls)), results in an _incoming message_ to actor A.

> A single message is executed _atomically_. This means that the code executed within one message either executes successfully or not at all. This also means that any _state mutations_ within a single message are either all committed or none of them are committed.

Reviewer: ```suggestion
A single message is executed _atomically_. This means that the code executed within one message either executes successfully or not at all. This also means that any _state changes_ within a single message are either all committed or none of them are committed. These state changes also include any messages sent with calls  to shared function. Calls are queued locally and only sent on successful commit.
```

Unless an [`Error`](#errors) is thrown _intentionally_ during execution, the order in which computations and state mutations occur is not relevant for the atomicity of a function. The whole function is executed as a single message that either fails or succeeds in its entirety.

### Shared functions that `await`

The [async-await](#async-and-await) example earlier looks simple, but there is a lot going on there. The function call is executed as several separate messages:

The first line `let future : async Nat = read()` is executed as part of the first message.

The second line `let result = await future;` keyword ends the first message and any state changes made during execution of the message are committed. The `await` also calls `read()` and suspends execution until a response is received.

The call to `read()` is executed as a separate message and could possibly be executed remotely in another actor, see [inter-canister calls](#inter-canister-calls). The message sent to `read()` could possibly result in several messages if the `read()` function also `await`s in its body.

> If the sent message executes successfully, a _callback_ is made to `call_read` that is executed as yet another message as the the last line `result + 1`.

Reviewer: ```suggestion
If the sent message executes successfully, a _callback_ is made to `call_read` that is executed as yet another message as the last line `result + 1`.
```

Reviewer: The callback writes the result into the future, completing it, and resumes the code that was waiting on the future, i.e. the last line `result +1`.

Examples of messaging restrictions:

- Canister installation can execute code (and possibly [trap](#traps)), but not send messages.

- A canister query method cannot send messages ([yet](#inter-canister-query-calls))

- The `await` and `await*` constructs are only allowed in an asynchronous context.

- The `throw` expression (to throw an [error](#errors)) is only allowed in an asynchronous context.

- The `try-catch` expression is only allowed in an asynchronous context. This is because error handling is supported for _messaging errors_ only and, like messaging itself, confined to asynchronous contexts.

[See official docs](https://internetcomputer.org/docs/current/motoko/main/actors-async/#messaging-restrictions)

### Message ordering

Messages in an actor are always executed sequentially, meaning one after the other and never in parallel. As a programmer, you have no control over the order in which _incoming messages_ are executed.

> You can only control the order in which you send messages to other actors, with the guarantee that they will be executed in the order you sent them. But you have no guarantee on the order in which you receive the _callbacks_ for those messages.

Reviewer: ```suggestion
You can only control the order in which you send messages to other actors, with the guarantee that, for a particular destination actor, they will be executed in the order you sent them. Messages sent to different actors may be executed out of order. 
You have no guarantee on the order in which you receive the results of any messages.
```

- A trap may occur intentionally for development purposes, see [Debug.trap()](/base-library/utils/debug.html)
- A trap can happen anywhere code runs in an actor, not only in an [asynchronous context](#messaging-restrictions).

```motoko
{{#include _mo/async-calls3-trap.mo:a}}
```

We import the [`Debug`](/base-library/utils/debug.html) module.

The shared function `incrementAndTrap` fails and does not return `()`. Instead it causes the execution to trap (see [Debug](/base-library/utils/debug.html) module).

**Both the first and second state mutation are NOT committed.**

After `incrementAndTrap`, our mutable variable `state` is not changed at all.

> **NOTE**  
> _Usually a trap occurs without `Debug.trap()` during execution of code, for example at underflow or overflow of [bounded types](/base-library/primitive-types/bounded-number-types.html) or other runtime failures, see [traps](#traps)._  
>  
> _[Assertions](/common-programming-concepts/assertions.html) also generate a trap if their argument evaluates to `false`_

> ## Async* and Await*

Reviewer: GOT HERE

```motoko
{{#include _mo/async-calls7.mo}}
```

`compute` is a [private function](/common-programming-concepts/functions.html#private-functions) with `async* Nat` return type. Calling it directly yields a _future_ of type `async* Nat` and resumes execution without blocking. This future needs to be awaited using `await*` for the computation to actually execute (unlike the case with 'ordinary' `async` [atomic functions that send messages](#atomic-functions-that-send-messages)).

`await*` also suspends execution, until a result is obtained. We could also pass around the future within our actor code and only `await*` it when we actually need the result.

We `await*` our function `compute` from within an [asynchronous context](#messaging-restrictions).

We `await*` our function `compute` from within an 'ordinary' shared `async` function `call_compute` or from within another private `async*` like `call_compute2`.

In the case of `call_compute` we obtain the result by first declaring a future and then `await*`ing it. In the case of `call_compute2` we `await*` the result directly.

> `compute` and `call_compute` are not part of the [_actor type_](/internet-computer-programming-concepts/actors.html#actor-type) and [_public API_](/internet-computer-programming-concepts/async-data/candid.html#actor-interfaces), because it is a [private function](/common-programming-concepts/functions.html#private-functions).

Reviewer: ```suggestion
`compute` and `call_compute` are not part of the [_actor type_](/internet-computer-programming-concepts/actors.html#actor-type) and [_public API_](/internet-computer-programming-concepts/async-data/candid.html#actor-interfaces), because they are [private functions](/common-programming-concepts/functions.html#private-functions).
```
